packages/cosmic-swingset/src/helpers/sync-xfer.js

```diff
@@ -0,0 +1,160 @@
+// @ts-check
+
+import { Far } from '@endo/far';
+
+// FIXME: for debugging the buffer splitting/rejoining:
+// const DEFAULT_BUFFER_SIZE = 12;
+const DEFAULT_BUFFER_SIZE = 64 * 1024;
+export { DEFAULT_BUFFER_SIZE };
+
+// const TRANSFER_UNSPECIFIED = 0;
+const TRANSFER_CLIENT_READY = 1;
+const TRANSFER_READY_TO_READ = 2;
+const TRANSFER_READ_PARTIAL = 3;
+const TRANSFER_READ_DONE = 4;
+
+const { Fail } = assert;
+
+const debug = (..._args) => {};
+
+/**
+ * @template T
+ * @typedef {Generator<undefined, T, Uint8Array | undefined>} Decoder<T>
+ */
+
+/**
+ * @typedef {Generator<Uint8Array | undefined, Uint8Array, Uint8Array>} Encoder<T>
+ */
+
+/**
+ * @returns {Decoder<string>}
+ */
+export function* makeTextDecoder() {
+  // console.error('created decoder!');
+  const td = new TextDecoder('utf-8');
+  let s = '';
+  for (;;) {
+    const buf = yield;
+    // console.error('got buf', buf);
+    if (buf === undefined) {
+      return s;
+    }
+    s += td.decode(buf, { stream: true });
+  }
+}
+harden(makeTextDecoder);
+
+/**
+ * @param {string} input
+ * @returns {Encoder}
+ */
+export function* makeTextEncoder(input) {
+  const te = new TextEncoder();
+
+  let buf = yield;
+
+  // TODO: Maybe encode partials directly into buf, not create a fresh one.
+  buf = te.encode(input);
+  return buf;
+}
+harden(makeTextEncoder);
+
+export const makeTransferClient = sharedBuf => {
+  const control = new Int32Array(sharedBuf, 0, 2);
+  const outb = new Uint8Array(control.buffer, 2 * Int32Array.BYTES_PER_ELEMENT);
+
+  return Far('TransferClient', {
+    prepareToRead(cb) {
+      control[0] = TRANSFER_CLIENT_READY;
+      cb(sharedBuf);
+      Atomics.wait(control, 0, TRANSFER_CLIENT_READY);
+      debug('current state', control[0]);
+    },
+    /**
+     * @template T
+     * @param {Decoder<T>} decoder
+     */
+    read(decoder) {
+      decoder.next();
+      Atomics.wait(control, 0, TRANSFER_READY_TO_READ);
+      while (control[0] === TRANSFER_READ_PARTIAL) {
+        const thisTransfer = Math.min(control[1], outb.byteLength);
```

Comment:
This should be an assertion.
```suggestion
        const thisTransfer = control[1];
        if (thisTransfer > outb.byteLength) {
          throw Error(`transfer count ${thisTransfer} exceeds buffer size ${outb.byteLength}`);
        }
```
...and more generally, I think all the TransferClient and TransferServer methods should assert the call-initial state of `control[0]` (which right now is only present in `prepareToWrite`).